### PR TITLE
Refactor `on_close_block`, take #2

### PR DESCRIPTION
**Pull request by Remagpie**

### CODE REVIEW

File: core/src/consensus/tendermint/engine.rs [hunk 1]
```diff
@@ -273,7 +267,7 @@ impl ConsensusEngine for Tendermint {
 
         let inner = self.inner.clone();
         let extension = service.register_extension(move |api| TendermintExtension::new(inner, timeouts, api));
-        let client = Weak::clone(self.client.read().as_ref().unwrap());
+        let client = Arc::downgrade(&self.client().unwrap());
```
Comment (foriequal0) [inline core/src/consensus/tendermint/engine.rs]: Note that this patch will changes the behavior (by small change)
The previous one panics only when the `self.client == None`, however, the changed one will panics 1. `self.client == None`, 2. `Weak::upgrade(self.client.unwrap()).is_none()`

File: core/src/consensus/tendermint/engine.rs [hunk 2]
```diff
@@ -137,12 +137,16 @@ impl ConsensusEngine for Tendermint {
     fn on_close_block(
         &self,
         block: &mut ExecutedBlock,
-        parent_header: &Header,
-        parent_common_params: &CommonParams,
         term_common_params: Option<&CommonParams>,
```
Comment (foriequal0) [inline core/src/consensus/tendermint/engine.rs]: Can you remove this parameter too?

Comment (Remagpie) [reply core/src/consensus/tendermint/engine.rs]: I tried to do that, and I thought it'd be a bit complex for now.
How about removing this after https://github.com/CodeChain-io/codechain/pull/1895 is merged?

File: core/src/consensus/tendermint/engine.rs (hunk 1 above)
Comment (Remagpie) [reply core/src/consensus/tendermint/engine.rs]: Well, it does change the behavior, but I don't think it matters.
`register_network_extension_to_service` is called only once at the initialization, and I think there's no possibility of panic in that situation.

**Pull request merged**